qiskit/algorithms/linear_solvers/matrices/numpy_matrix.py

```diff
@@ -183,6 +183,12 @@ def _reset_registers(self, num_state_qubits: int) -> None:
         self.qregs = [qr_state]
         self._qubits = qr_state[:]
 
+        from qiskit.circuit.quantumcircuit import BitLocations
+
+        self._qubit_indices = {bit: BitLocations(idx, []) for idx, bit in enumerate(self._qubits)}
+        for reg_idx, bit in enumerate(qr_state):
+            self._qubit_indices[bit][1].append((qr_state, reg_idx))
+
```

Comment:
Is it worth delegating this to `self.add_register(qr_state)`, and similar elsewhere in this PR?  If not, the import at least should be able to safely go at the top of the file, I think?

Also: here, and elsewhere, we're using indexed access like `BitLocations(...)[1]` instead of `BitLocations(...).registers`.  There's a small time penalty for attribute access in Python 3.7, but in 3.8+ they're the same.  It's probably worth using attribute access everywhere for readability - if we're not using the attributes for speed reasons, then a regular tuple might be better.

Reply:
> Is it worth delegating this to `self.add_register(qr_state)`, and similar elsewhere in this PR? If not, the import at least should be able to safely go at the top of the file, I think?

We could have something like `self._qubits = []; self._qregs = []; self._ancillas = []; self._qubit_indices = {}; self.add_register(qr_state);` which is definitely clearer. I'll update the PR in that direction.

> Also: here, and elsewhere, we're using indexed access like `BitLocations(...)[1]` instead of `BitLocations(...).registers`. There's a small time penalty for attribute access in Python 3.7, but in 3.8+ they're the same. It's probably worth using attribute access everywhere for readability - if we're not using the attributes for speed reasons, then a regular tuple might be better.

Good point, I'll remove the index access in favor of attributes.